docs/DesignDocs/PluggableAllocators.md

```diff
@@ -0,0 +1,142 @@
+  (e.g [hello_worlf.conf](https://github.com/openenclave/openenclave/blob/e2e440ec134d37e107b5a296fdc508a3a643598e/samples/helloworld/enclave/helloworld.conf#L6)) </br>
+  `oecore` may reserve a small portion for its own use and pass the rest to this replacement function. The allocator is expected to perform any global initialization in this function. </br>
+  This function is called prior to any global initialization.
+
+
+- `void oe_allocator_cleanup(void)` </br>
+  This function will be called by `oecore` at an appropriate time during enclave termination. `atexit` functions will be executed prior to `oe_allocator_cleanup`. </br>
+  `oe_allocator_cleanup` may not register functions with `atexit`.
+
+
+- `void oe_allocator_thread_init(void)` </br>
+  This function will be called by `oecore` during initialization of an enclave thread, after the thread-local variables for the thread have been initialized. </br>
+  In the future, if `oecore` needs to hold onto any thread-specific data on behalf of the allocator, it can be returned in this function. Currently the recommendation is to use `__thread` variables to store thread-specific allocator data.
+
+
+- `void oe_allocator_thread_cleanup(void)` </br>
+  This function will be called by `oecore` just prior to the termination of enclave thread. Thread-specific exit functions (`__cxa_thread_atexit` functions) will be executed prior to `oe_allocator_thread_cleanup`. </br>
+  `oe_allocator_thread_cleanup` may not use `__cxa_thread_atexit`; that is it cannot access C++ thread-local variables with non-trivial destructors for the first time. </br>
+  It is important that the allocators be aware of the nature of enclave threads, and be designed accordingly. Enclave threads can be short-lived. For OE SGX, enclave threads last only for the duration of an ecall. </br>
+  Additionally, the same thread-control structure (`sgx_tcs_t`) can be bound to different host threads at different times during the lifespan of an enclave.
+
+
+- `void* oe_allocator_malloc(size_t size)` </br>
+  This function will be called by `oecore` to implement `malloc`. Depending upon the build settings, the allocated memory might be tracked by the `debug-malloc` feature.
+
+
+- `void oe_allocator_free(void* ptr)` </br>
+  This function will be called by `oecore` to implement `free`.
+
+
+- `void* oe_allocator_calloc(size_t nmemb, size_t size)` </br>
+  This function will be called by `oecore` to implement `calloc`.
+
+
+- `void* oe_allocator_realloc(void* ptr, size_t size)` </br>
+  This function will be called by `oecore` to implement `realloc`.
+
+
+- `void oe_allocator_posix_memalign(void** memptr, size_t alignment, size_t size)` </br>
+  This function will be called by `oecore` to implement [posix_memalign](https://linux.die.net/man/3/posix_memalign). Memory allocated by this function can be freed via `free`.
+
+- `void oe_allocator_reusable_size(void* ptr)` </br>
```

Comment:
> Can you elaborate on what this is supposed to do? Why "reusable size" instead of just having the allocator implement `oe_allocator_malloc_usable_size()` directly?
> 
That ought to have been `oe_allocator_malloc_size`. Fixed.

Reply:
I added the following lines to the doc. 
"The set of replacement functions are mostly equivalents of the functions required by the GNU C Library for plugging-in custom allocators. Additionally, there are replacement functions that allow the allocator to work well with enclave threads."